### PR TITLE
Added StringCommands Decr, DecrBy, Incr, IncrBy, IncrByFloat, Mget and Mset and Refactored integration tests so there is no duplication of tests for the baseClient interface.

### PR DESCRIPTION
**Pull request by SanHalacogluImproving**
-Added StringCommands Decr, DecrBy, Incr, IncrBy, IncrByFloat, Mget and Mset.

-Added integration and unit tests for the new commands.

-Refactored integration tests so there is no duplication of tests for the baseClient interface.

### CODE REVIEW

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 1]
```diff
@@ -123,6 +132,42 @@ protected String handleStringResponse(Response response) {
...
+     * @return Response as an <code>Long</code>
+     */
+    protected Long handleLongResponse(Response response) {
+        Object value = handleObjectResponse(response);
+        if (value instanceof Long) {
+            return (Long) value;
+        }
+        throw new RedisException(
+                "Unexpected return type from Redis: got "
+                        + value.getClass().getSimpleName()
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/BaseClient.java:149]: `value` could be `null`. Add a protection:
https://github.com/Bit-Quill/glide-for-redis/blob/42b0498da2e1ef89947cf529d4e56b0893b91904/java/client/src/main/java/glide/api/BaseClient.java#L148

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 2]
```diff
@@ -123,6 +132,42 @@ protected String handleStringResponse(Response response) {
...
+     * @return Response as an <code>Double</code>
+     */
+    protected Double handleDoubleResponse(Response response) {
+        Object value = handleObjectResponse(response);
+        if (value instanceof Double) {
+            return (Double) value;
+        }
+        throw new RedisException(
+                "Unexpected return type from Redis: got "
+                        + value.getClass().getSimpleName()
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/BaseClient.java:167]: and here

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 3]
```diff
@@ -187,4 +232,54 @@ public CompletableFuture<String> set(String key, String value, SetOptions option
...
+    public CompletableFuture<String[]> mget(String[] keys) {
+        return commandManager
+                .submitNewCommand(MGet, keys, this::handleArrayResponse)
+                .thenApply(
+                        objectsArray ->
+                                Arrays.stream(objectsArray).map(object -> (String) object).toArray(String[]::new));
+    }
+
+    @Override
+    public CompletableFuture<String> mset(HashMap<String, String> keyValueMap) {
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/BaseClient.java:274]:
```suggestion
    public CompletableFuture<String> mset(Map<String, String> keyValueMap) {
```

File: java/client/src/main/java/glide/api/commands/StringCommands.java [hunk 4]
```diff
@@ -47,4 +48,90 @@ public interface StringCommands {
...
+    CompletableFuture<String[]> mget(String[] keys);
+
+    /**
+     * Set multiple keys to multiple values in a single operation.
+     *
+     * @see <a href="https://redis.io/commands/mset/">redis.io</a> for details.
+     * @param keyValueMap A key-value map consisting of keys and their respective values to set.
+     * @return Always "Ok".
+     */
+    CompletableFuture<String> mset(HashMap<String, String> keyValueMap);
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StringCommands.java:136]:
```suggestion
    CompletableFuture<String> mset(Map<String, String> keyValueMap);
```

File: java/integTest/src/test/java/glide/BaseClientTests.java [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+    @Test
+    @SneakyThrows
+    public void decr_and_decrBy_existing_key() {
+        String key = RandomStringUtils.randomAlphabetic(10);
+
+        for (BaseClient client : clients) {
+            assertEquals(OK, client.set(key, "10").get(10, SECONDS));
+
+            assertEquals(client.decr(key).get(10, SECONDS), 9);
+            assertEquals(client.get(key).get(10, SECONDS), "9");
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/BaseClientTests.java:298]: expected value is on the left side in all asserts, please reorder all of them.

File: java/integTest/src/test/java/glide/BaseClientTests.java [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+        String nonExisting = RandomStringUtils.randomAlphabetic(10);
+        String value = RandomStringUtils.randomAlphabetic(10);
+        HashMap<String, String> keyValueMap =
+                new HashMap<>() {
+                    {
+                        put(key1, value);
+                        put(key2, value);
+                        put(key3, value);
+                    }
+                };
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/BaseClientTests.java:407]:
```suggestion
        Map<String, String> keyValueMap = Map.of(key1, value, key2, value, key3, value);
```

File: java/integTest/src/test/java/glide/BaseClientTests.java [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+            assertTrue(incrByException.getCause() instanceof RequestException);
+            assertTrue(incrByException.getCause().getMessage().contains("value is not an integer"));
+
+            Exception incrByFloatException =
+                    assertThrows(
+                            ExecutionException.class, () -> client.incrByFloat(key1, 3.5).get(10, SECONDS));
+
+            assertTrue(incrByFloatException.getCause() instanceof RequestException);
+            assertTrue(
+                    incrByFloatException.getCause().getMessage().contains("value is not a valid float"));
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/BaseClientTests.java:388]: test `incr*` commands with missing key too
Consider testing with `Long.MAX_VALUE` too (and max double)

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 8]
```diff
@@ -162,20 +153,6 @@ public void info_with_routing_and_options() {
         }
     }
 
-    @Test
-    @SneakyThrows
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:166]: why?

File: java/integTest/src/test/java/glide/standalone/CommandTests.java [hunk 9]
```diff
@@ -90,57 +82,4 @@ public void info_with_everything_option() {
             assertTrue(data.contains("# " + section), "Section " + section + " is missing");
         }
     }
-
-    @Test
-    @SneakyThrows
-    public void set_and_get_without_options() {
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/standalone/CommandTests.java:96]: why?

**Review by Yury-Fridlyand — COMMENTED** [review]
~Please add implementation for cluster client too~

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 10]
```diff
@@ -123,6 +132,42 @@ protected String handleStringResponse(Response response) {
...
+     * @param response Redis protobuf message
+     * @return Response as an <code>Double</code>
+     */
+    protected Double handleDoubleResponse(Response response) {
+        Object value = handleObjectResponse(response);
+        if (value instanceof Double) {
+            return (Double) value;
+        }
+        throw new RedisException(
+                "Unexpected return type from Redis: got "
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/BaseClient.java:166]: We should consider nulls here: 
https://github.com/aws/glide-for-redis/blob/42b0498da2e1ef89947cf529d4e56b0893b91904/java/client/src/main/java/glide/api/BaseClient.java#L148

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 11]
```diff
@@ -123,6 +132,42 @@ protected String handleStringResponse(Response response) {
...
+     * @param response Redis protobuf message
+     * @return Response as an <code>Long</code>
+     */
+    protected Long handleLongResponse(Response response) {
+        Object value = handleObjectResponse(response);
+        if (value instanceof Long) {
+            return (Long) value;
+        }
+        throw new RedisException(
+                "Unexpected return type from Redis: got "
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/BaseClient.java:148]: We should consider nulls here:
https://github.com/aws/glide-for-redis/blob/42b0498da2e1ef89947cf529d4e56b0893b91904/java/client/src/main/java/glide/api/BaseClient.java#L148

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 12]
```diff
@@ -123,6 +132,42 @@ protected String handleStringResponse(Response response) {
...
+                        + value.getClass().getSimpleName()
+                        + " expected Long");
+    }
+
+    /**
+     * Extracts the response value from the Redis response and either throws an exception or returns
+     * the value as an <code>Double</code>.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/BaseClient.java:155]:
```suggestion
     * the value as a <code>Double</code>.
```

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 13]
```diff
@@ -123,6 +132,42 @@ protected String handleStringResponse(Response response) {
                         + " expected String");
     }
 
+    /**
+     * Extracts the response value from the Redis response and either throws an exception or returns
+     * the value as an <code>Long</code>.
+     *
+     * @param response Redis protobuf message
+     * @return Response as an <code>Long</code>
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/BaseClient.java:140]:
```suggestion
     * @return Response as a <code>Long</code>
```

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 14]
```diff
@@ -123,6 +132,42 @@ protected String handleStringResponse(Response response) {
                         + " expected String");
     }
 
+    /**
+     * Extracts the response value from the Redis response and either throws an exception or returns
+     * the value as an <code>Long</code>.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/BaseClient.java:137]:
```suggestion
     * the value as a <code>Long</code>.
```

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 15]
```diff
@@ -123,6 +132,42 @@ protected String handleStringResponse(Response response) {
...
+                        + value.getClass().getSimpleName()
+                        + " expected Long");
+    }
+
+    /**
+     * Extracts the response value from the Redis response and either throws an exception or returns
+     * the value as an <code>Double</code>.
+     *
+     * @param response Redis protobuf message
+     * @return Response as an <code>Double</code>
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/BaseClient.java:158]:
```suggestion
     * @return Response as a <code>Double</code>
```

File: java/client/src/main/java/glide/api/commands/StringCommands.java [hunk 16]
```diff
@@ -47,4 +48,90 @@ public interface StringCommands {
      *     Otherwise, return <code>OK</code>.
      */
     CompletableFuture<String> set(String key, String value, SetOptions options);
+
+    /**
+     * Increment the string representing a floating point number stored at <em>key</em> by
+     * <em>amount</em>. By using a negative increment value, the result is that the value stored at
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/StringCommands.java:54]: I'd prefer we use `<code></code>` tags instead of `<em>`.  
```suggestion
     * <code>amount</code>. By using a negative increment value, the result is that the value stored at
```

File: java/client/src/main/java/glide/api/commands/StringCommands.java [hunk 17]
```diff
@@ -47,4 +48,90 @@ public interface StringCommands {
...
+     * <em>key</em> is decremented. If <em>key</em> does not exist, it is set to 0 before performing
+     * the operation.
+     *
+     * @see <a href="https://redis.io/commands/decr/">redis.io</a> for details.
+     * @param key The key to increment its value.
+     * @return The value of <em>key</em> after the increment. An error is raised if <em>key</em>
+     *     contains a value of the wrong type, or the current key content is not parsable as a double
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/StringCommands.java:61]:
```suggestion
     *     contains a value of the wrong type, or the current <code>key</code> content is not parsable as a double
```

File: java/client/src/main/java/glide/api/commands/StringCommands.java [hunk 18]
```diff
@@ -47,4 +48,90 @@ public interface StringCommands {
...
+     * @param key The key to increment its value.
+     * @param amount The amount to increment.
+     * @return The value of <em>key</em> after the increment. An error is raised if <em>key</em>
+     *     contains a value of the wrong type, or the current key content is not parsable as a double
+     *     precision floating point number.
+     */
+    CompletableFuture<Double> incrByFloat(String key, double amount);
+
+    /**
+     * Retrieve the values of multiple keys.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/StringCommands.java:120]:
```suggestion
     * Retrieve the values of multiple <code>keys</code>.
```

File: java/client/src/main/java/glide/api/commands/StringCommands.java [hunk 19]
```diff
@@ -47,4 +48,90 @@ public interface StringCommands {
...
+     *
+     * @see <a href="https://redis.io/commands/mget/">redis.io</a> for details.
+     * @param keys A list of keys to retrieve values for.
+     * @return A list of values corresponding to the provided keys. If a key is not found, its
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/StringCommands.java:124]:
```suggestion
     * @return A list of values corresponding to the provided <code>keys</code>. If a <code>key</code> is not found, its
```

File: java/client/src/main/java/glide/api/commands/StringCommands.java [hunk 20]
```diff
@@ -47,4 +48,90 @@ public interface StringCommands {
...
+     *     corresponding value in the list will be null.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/StringCommands.java:125]:
```suggestion
     *     corresponding value in the list will be <code>null</code>.
```

File: java/client/src/main/java/glide/api/commands/StringCommands.java [hunk 21]
```diff
@@ -47,4 +48,90 @@ public interface StringCommands {
...
+     */
+    CompletableFuture<String[]> mget(String[] keys);
+
+    /**
+     * Set multiple keys to multiple values in a single operation.
+     *
+     * @see <a href="https://redis.io/commands/mset/">redis.io</a> for details.
+     * @param keyValueMap A key-value map consisting of keys and their respective values to set.
+     * @return Always "Ok".
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/StringCommands.java:134]:
```suggestion
     * @return Always <code>"Ok"</code>.
```

File: java/integTest/src/test/java/glide/BaseClientTests.java [hunk 22]
```diff
@@ -0,0 +1,416 @@
...
+    @AfterAll
+    @SneakyThrows
+    public static void teardown() {
+        standaloneClient.close();
+        clusterClient.close();
+    }
+
+    @Test
+    @SneakyThrows
+    public void ping() {
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/BaseClientTests.java:78]: they may ask you to make this change in a separate PR

File: java/integTest/src/test/java/glide/BaseClientTests.java [hunk 23]
```diff
@@ -0,0 +1,416 @@
...
+        for (BaseClient client : clients) {
+            var data = client.ping().get(10, SECONDS);
+            assertEquals("PONG", data);
+        }
+    }
+
+    @Test
+    @SneakyThrows
+    public void ping_with_message() {
+        for (BaseClient client : clients) {
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/BaseClientTests.java:88]: you should consider making this a @ParameterizedTest
and pass in the pair of clients as arguments. 

Like: 
```
private List<Arguments> provideClients() {
        return List.of(
            Arguments.of(standaloneClient),
            Arguments.of(clusterClient)
        );
    }

    @ParameterizedTest
    @MethodSource("provideClients")
    public void test(BaseClient client) {
            var data = client.ping("H3LL0").get(10, SECONDS);
            assertEquals("H3LL0", data);
     }
```

File: java/integTest/src/test/java/glide/BaseClientTests.java [hunk 24]
```diff
@@ -0,0 +1,416 @@
...
+            assertNull(client.get(key2).get(10, SECONDS));
+            assertEquals(client.decrBy(key2, 3).get(10, SECONDS), -3);
+            assertEquals(client.get(key2).get(10, SECONDS), "-3");
+        }
+    }
+
+    @Test
+    @SneakyThrows
+    public void incr_commands_existing_key() {
+        String key = RandomStringUtils.randomAlphabetic(10);
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/BaseClientTests.java:325]: ❤️

File: java/integTest/src/test/java/glide/BaseClientTests.java [hunk 25]
```diff
@@ -0,0 +1,416 @@
...
+            assertEquals(client.get(key).get(10, SECONDS), "20.5");
+        }
+    }
+
+    @Test
+    @SneakyThrows
+    public void incr_commands_non_existing_key() {
+        String key1 = RandomStringUtils.randomAlphabetic(10);
+        String key2 = RandomStringUtils.randomAlphabetic(10);
+        String key3 = RandomStringUtils.randomAlphabetic(10);
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/BaseClientTests.java:346]: you randomize the key, but not the increment value?

File: java/integTest/src/test/java/glide/BaseClientTests.java [hunk 26]
```diff
@@ -0,0 +1,416 @@
...
+        }
+    }
+
+    @Test
+    @SneakyThrows
+    public void incr_commands_existing_key() {
+        String key = RandomStringUtils.randomAlphabetic(10);
+
+        for (BaseClient client : clients) {
+            assertEquals(OK, client.set(key, "10").get(10, SECONDS));
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/BaseClientTests.java:328]: you randomize the key, but not the increment value?

File: java/integTest/src/test/java/glide/BaseClientTests.java [hunk 27]
```diff
@@ -0,0 +1,416 @@
...
+
+    @Test
+    @SneakyThrows
+    public void decr_and_decrBy_existing_key() {
+        String key = RandomStringUtils.randomAlphabetic(10);
+
+        for (BaseClient client : clients) {
+            assertEquals(OK, client.set(key, "10").get(10, SECONDS));
+
+            assertEquals(client.decr(key).get(10, SECONDS), 9);
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/BaseClientTests.java:297]: test with more than just 10 - maybe also `0` and `-10`?

File: java/integTest/src/test/java/glide/BaseClientTests.java (hunk 22 above)
Comment (SanHalacogluImproving) [reply java/integTest/src/test/java/glide/BaseClientTests.java:78]: Probably that's what I am thinking.

File: java/integTest/src/test/java/glide/BaseClientTests.java (hunk 23 above)
Comment (SanHalacogluImproving) [reply java/integTest/src/test/java/glide/BaseClientTests.java:88]: Will check it out.

File: java/integTest/src/test/java/glide/BaseClientTests.java (hunk 27 above)
Comment (SanHalacogluImproving) [reply java/integTest/src/test/java/glide/BaseClientTests.java:297]: Took these from the python and node IT tests.

File: java/integTest/src/test/java/glide/BaseClientTests.java (hunk 5 above)
Comment (SanHalacogluImproving) [reply java/integTest/src/test/java/glide/BaseClientTests.java:298]: will fix it

File: java/integTest/src/test/java/glide/BaseClientTests.java (hunk 26 above)
Comment (SanHalacogluImproving) [reply java/integTest/src/test/java/glide/BaseClientTests.java:328]: Took these from the Python and node IT tests.

File: java/integTest/src/test/java/glide/BaseClientTests.java (hunk 25 above)
Comment (SanHalacogluImproving) [reply java/integTest/src/test/java/glide/BaseClientTests.java:346]: Took these from the Python and node IT tests.

File: java/integTest/src/test/java/glide/BaseClientTests.java (hunk 7 above)
Comment (SanHalacogluImproving) [reply java/integTest/src/test/java/glide/BaseClientTests.java:388]: Are you referring to null key value or not existing in the DB? I can add the maxes I copied over all the IT tests from python and node.

File: java/integTest/src/test/java/glide/standalone/CommandTests.java (hunk 9 above)
Comment (SanHalacogluImproving) [reply java/integTest/src/test/java/glide/standalone/CommandTests.java:96]: Why what? Sneaky throws or why did I remove them? If it is the latter these are tests meant to be in baseClient tests not only the cluster I transfered them to there.

File: java/integTest/src/test/java/glide/BaseClientTests.java (hunk 7 above)
Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/BaseClientTests.java:388]: Try both - null and missing.
Re: max value - you will get an error on Redis side, rust core lib or in java. Good to know and record the limitations.

Comment (SanHalacogluImproving) [reply java/integTest/src/test/java/glide/BaseClientTests.java:388]: Fair but does that mean we need to write out the same test for all functions that incr decr by a number? There multiple variations for hash_commands etc...

Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/BaseClientTests.java:388]: Well, yes. If we have the same error for null, missing and max values - you can combine those into a single paramtrized test.

File: java/integTest/src/test/java/glide/BaseClientTests.java (hunk 22 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/BaseClientTests.java:78]: Good news.  I've moved all the ping, & get/set calls to their upstream PRs.  I ended up calling the test file SharedCommandTests - like they did for Node tests.  
* https://github.com/aws/glide-for-redis/pull/919/files#diff-5db9a3bd41d647cfdba521ea248ea07d7c78bc9aaa41a26bb923d4f92a3efdd2
* https://github.com/aws/glide-for-redis/pull/914/files#diff-5db9a3bd41d647cfdba521ea248ea07d7c78bc9aaa41a26bb923d4f92a3efdd2
